### PR TITLE
Bug 1904584: Check sa owner 4.5

### PR DESCRIPTION
**Pull request by dinhxuanvu**
**Description of the change:**

**Motivation for the change:**

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: test/e2e/subscription_e2e_test.go [hunk 1]
```diff
@@ -1346,8 +1350,10 @@ var _ = Describe("Subscription", func() {
 		packageName3 := genName("cpackage")
 
 		namedStrategy := newNginxInstallStrategy((genName("dep")), permissions, nil)
-		depNamedStrategy := newNginxInstallStrategy((genName("dep")), permissions, nil)
-		depNamedStrategy2 := newNginxInstallStrategy((genName("dep")), permissions, nil)
+		depNamedStrategy := newNginxInstallStrategy((genName("dep")), permissions2, nil)
+		depNamedStrategy2 := newNginxInstallStrategy((genName("dep")), permissions3, nil)
+		depNamedStrategy3 := newNginxInstallStrategy((genName("dep")), permissions4, nil)
+		depNamedStrategy4 := newNginxInstallStrategy((genName("dep")), permissions5, nil)
```
Comment (njhale) [inline test/e2e/subscription_e2e_test.go:1356]: Is there an issue with shared ServiceAccounts? Why wasn't this an issue in 4.5 and 4.6?

File: test/e2e/subscription_e2e_test.go [hunk 2]
```diff
@@ -1300,6 +1300,10 @@ var _ = Describe("Subscription", func() {
 		crClient := newCRClient()
 
 		permissions := deploymentPermissions()
+		permissions2 := deploymentPermissions()
```
Comment (njhale) [inline test/e2e/subscription_e2e_test.go:1303]: Is there an issue with shared ServiceAccounts? Why wasn't this an issue in 4.5 and 4.6?

**Comment by njhale** [comment]
/hold

I'm not sure I understand the necessity of https://github.com/operator-framework/operator-lifecycle-manager/pull/1909/commits/21936f86e98ec398edd016a6e483d43c7387d008 -- would like a walk through before we diverge from the original patch.

**Comment by dinhxuanvu** [comment]
@njhale I have identified the diff between 4.5 and 4.6. The two test cases in questions were slightly changed in 4.6. The install strategy doesn't have permissions in 4.6 which leads to no SA created. I have updates the test cases in 4.5 to no longer have permission anymore who they align with the test cases in 4.6.

**Comment by njhale** [comment]
> The install strategy doesn't have permissions in 4.6 which leads to no SA created. I have updates the test cases in 4.5 to no longer have permission anymore who they align with the test cases in 4.6.

Was the test case in 4.5 failing because permissions were present? If so, that would be a regression unless the test is somehow incorrect; i.e. operators in an upgrade chain should be able to share serviceaccounts.

If this test was not failing, then I don't think we should be refactoring tests in a backport.

**Comment by njhale** [comment]
/retest

It looks like our upstream e2e tests pass consistently. 

I think we want a running catalog of our flaky downstream tests. We need to get a handle on tests that flake because:
- the test is non-eventual/brittle
- controllers are not fully level-triggered

If we weed out all of the former, the latter should become apparent.

**Comment by dinhxuanvu** [comment]
> > The install strategy doesn't have permissions in 4.6 which leads to no SA created. I have updates the test cases in 4.5 to no longer have permission anymore who they align with the test cases in 4.6.
> 
> Was the test case in 4.5 failing because permissions were present? If so, that would be a regression unless the test is somehow incorrect; i.e. operators in an upgrade chain should be able to share serviceaccounts.
> 
> If this test was not failing, then I don't think we should be refactoring tests in a backport.

Yes it failed because of permission presence. There is not an issue with ServiceAccount usage as it can shared along multiple operators. The issue here is the ownerref. The SA only has one ownerref as OLM doesn't currently aggregate ownerrefs. The SA is present to use but it only has one owner which often is the latest one that is being installed in InstallPlan step order. The SA owner check that is introduced in this PR will because of it. We can address this by opening another PR/BZ to aggregate ownerrefs.
Plus, the test cases in question don't need those permissions given that is not what the test cases are testing for. Hence, that's why they were removed in 4.6 I believe. It helped to reduce the amount of installed resources.

**Comment by awgreene** [comment]
> [APPROVALNOTIFIER] This PR is **NOT APPROVED**
> 
> This pull-request has been approved by: _[awgreene](https://github.com/operator-framework/operator-lifecycle-manager/pull/1909#issuecomment-756312443)_, _[dinhxuanvu](https://github.com/operator-framework/operator-lifecycle-manager/pull/1909#)_
> To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **njhale** after the PR has been reviewed.
> You can assign the PR to them by writing `/assign @njhale` in a comment when ready.
> 
> The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=operator-framework%2Foperator-lifecycle-manager).
> 
> Needs approval from an approver in each of these files:
> * **[OWNERS](https://github.com/operator-framework/operator-lifecycle-manager/blob/release-4.5/OWNERS)**
> 
> Approvers can indicate their approval by writing `/approve` in a comment
> Approvers can cancel approval by writing `/approve cancel` in a comment

I do not have approval on the 4.5 branch, applying an LGTM instead.

/lgtm

**Comment by dinhxuanvu** [comment]
@njhale I wonder if you have seen my response. I would like to move forward with this PR unless there is something else to discuss?

**Comment by benluddy** [comment]
> @njhale I wonder if you have seen my response. I would like to move forward with this PR unless there is something else to discuss?

The subscription tests that fail with this change exist to verify the results of dependency resolution. As they exist in 4.5, they are overspecified. In 4.6, the same tests now verify expected behaviors by inspecting the InstallPlan that is produced as an output of resolution. I would support removing the additional (excessive) CSV verification in 4.5 as well.

File: test/e2e/subscription_e2e_test.go [hunk 3]
```diff
@@ -1436,22 +1434,17 @@ var _ = Describe("Subscription", func() {
...
+		var got []string
+		Eventually(func() []string {
+			ip, err := crClient.OperatorsV1alpha1().InstallPlans(testNamespace).Get(context.TODO(), subscription.Status.InstallPlanRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil
+			}
+			got = ip.Spec.ClusterServiceVersionNames
+			return got
+		}).ShouldNot(BeNil())
+		require.ElementsMatch(GinkgoT(), []string{"nginx-a", "nginx-b-dep", "nginx-e-dep"}, got)
```
Comment (benluddy) [inline test/e2e/subscription_e2e_test.go:1447]: Nit: Mixture of Gomega matchers and testify assertions, and `Eventually` swallows any errors from the Get.

```suggestion
		Eventually(func() ([]string, error) {
			ip, err := crClient.OperatorsV1alpha1().InstallPlans(testNamespace).Get(context.TODO(), subscription.Status.InstallPlanRef.Name, metav1.GetOptions{})
			if err != nil {
				return nil, err
			}
			return ip.Spec.ClusterServiceVersionNames, nil
		}).Should(ConsistOf("nginx-a", "nginx-b-dep", "nginx-e-dep"))
```

**Comment by ashcrow** [comment]
https://github.com/operator-framework/operator-lifecycle-manager/pull/1890

**Comment by ashcrow** [comment]
Removed patch manager approval as this PR is consistently failing a required test.

**Comment by dinhxuanvu** [comment]
@operator-framework/patch-managers PTAL

**Comment by crawford** [comment]
@dinhxuanvu what's going on with the failing CI? I see @ashcrow's message about failing tests and then twenty `/retests`.

**Comment by dinhxuanvu** [comment]
@crawford Hey Alex. This PR passed CI before Stephen approved but after he did, it triggered the tests to rerun and somehow hit the rate limit issue and other infra issues (like this [one](https://storage.googleapis.com/origin-ci-test/pr-logs/pull/operator-framework_operator-lifecycle-manager/1909/pull-ci-operator-framework-operator-lifecycle-manager-release-4.5-e2e-aws-olm/1350142999744483328/build-log.txt)). It finally got cleaned-up and passed. I admit the test suit in this branch is a bit flaky.

**Comment by crawford** [comment]
Cherry-pick-approved removed. Too flaky for my taste and we already have too many fixes to be verified in this round.

**Comment by dinhxuanvu** [comment]
@bparees I'm not sure if it is too late for this week cherry-pick approval window but please take a look. Thanks.

**Pull request merged**